### PR TITLE
feature/TECH 521 simplified remote artifact caching

### PR DESCRIPTION
**Pull request by SamTheisens**
----
### 📕 [TECH-521](https://vandebron.atlassian.net/browse/TECH-521) Improve caching of artifacts jorgpost@vandebron.nl 
Tagged images should only be built once

----

Create a repo where we commit `.mpl` folders

* should be an optional feature

Define an optional part in the mpyl_config that specifies the “build artifact repository”. If that is specified we push and pull the `.mpl` folders

Figure out how to enable it:

* Add two subcommands to push and pull or 
`mpyl build artifacts pull`
`mpyl build run`
`mpyl build artifacts push`
* Add a flag to the build command to turn it off
`mpyl build run --cached`
`mpyl build run` → cached locally

**Workflow**

- After the first build commit the folder to the repo after deployment
#* on a feature branch, create a pr-XXXX (on a tag build create a branch with the tag name) branch in the artifact repo and push the `.mpl` folders there keeping the same folder structure
- On a consecutive build pull the files into a temporary folder, then do a recursive copy of the folder

Figure out a cleanup mechanism

* locally we have `mpyl projects clean`
* in repo: 

🏗️ Build [21](https://jenkins.k8s-serv-backend.vdbinfra.nl/job/MPyL%20Pipeline%20-%20Test/job/PR-249/21/display/redirect) ✅ Successful, started by _Jorg Post_  
🚀 *[cloudfront-service](https://cloudfront-service-249.test.nl/swagger/index.html)*, *example-dagster-user-code*, *job*, *kong-sync*, *[nodeservice](https://nodeservice-249.test.nl/swagger/index.html)*, *ocpp-first*, *ocpp-second*, *[sbtservice](https://sbtservice-249.test.nl/swagger/index.html)*, *sparkJob*  

[TECH-521]: https://vandebron.atlassian.net/browse/TECH-521?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/mpyl/artifacts/build_artifacts.py [hunk 1]
```diff
@@ -1,85 +1,62 @@
...
-            self.artifact_repo.push()
+        with TemporaryDirectory() as tmp_repo_dir:
+            with Repository.from_clone(
+                config=self.artifact_repo_config, repo_path=Path(tmp_repo_dir)
+            ) as artifact_repo:
+                if artifact_repo.does_branch_exist(branch_name=branch):
+                    self.logger.info(f"Fetching branch '{branch}' from remote")
+                    artifact_repo.checkout_branch(branch_name=branch)
+                self.logger.info(f"Branch {branch} does not exist in remote")
+                artifact_repo.
```
Comment (Jorg88) [inline src/mpyl/artifacts/build_artifacts.py:33]: Forgot to delete this?

File: src/mpyl/artifacts/build_artifacts.py [hunk 2]
```diff
@@ -1,85 +1,62 @@
...
+                config=self.artifact_repo_config, repo_path=repo_path
+            ) as artifact_repo:
+                branch_exists = artifact_repo.does_branch_exist(branch_name=branch)
+                if branch_exists:
+                    self.logger.info(f"Fetching branch '{branch}' from remote")
+                    artifact_repo.checkout_branch(branch_name=branch)
+                else:
+                    artifact_repo.create_branch(branch_name=branch)
+
+                shutil.rmtree(repo_path, ignore_errors=True)
```
Comment (Jorg88) [inline src/mpyl/artifacts/build_artifacts.py:48]: Won't the pr now include the deletion of the readme and gitignore files that are in main? I guess it's not the biggest deal for our purposes, but that's what I tried to avoid with the extra 'cache' folder

File: src/mpyl/artifacts/build_artifacts.py [hunk 3]
```diff
@@ -1,85 +1,62 @@
...
-            self.artifact_repo.stage(CACHE_FOLDER_NAME)
-            self.artifact_repo.commit(f"Add artifacts for {branch}")
-            self.artifact_repo.push()
+        with TemporaryDirectory() as tmp_repo_dir:
+            with Repository.from_clone(
+                config=self.artifact_repo_config, repo_path=Path(tmp_repo_dir)
+            ) as artifact_repo:
+                if artifact_repo.does_branch_exist(branch_name=branch):
+                    self.logger.info(f"Fetching branch '{branch}' from remote")
+                    artifact_repo.checkout_branch(branch_name=branch)
```
Comment (Jorg88) [inline src/mpyl/artifacts/build_artifacts.py:31]: We have to copy the files out of the repo here as well if you want to use `with TemporaryDirectory`

File: src/mpyl/artifacts/build_artifacts.py [hunk 4]
```diff
@@ -1,85 +1,62 @@
...
+                shutil.rmtree(repo_path, ignore_errors=True)
+                for file_path in file_paths:
+                    shutil.copytree(
+                        src=file_path,
+                        dst=repo_path / file_path.name,
+                        dirs_exist_ok=True,
+                    )
+
+                git = artifact_repo._repo.git
+                git.add(".")
```
Comment (Jorg88) [inline src/mpyl/artifacts/build_artifacts.py:57]: What's wrong with using the Repo functions itself like I did above?

File: src/mpyl/utilities/repo/__init__.py [hunk 5]
```diff
@@ -315,15 +313,8 @@ def push(self):
...
-    def does_branch_exist(self, branch_name: str, remote=False) -> bool:
-        self._repo.remote().fetch()
-        found_branches = [
-            branch.strip(" ").strip("*")
-            for branch in self._repo.git.branch(
-                "--list", "-r" if remote else None
-            ).splitlines()
-        ]
-        return (f"origin/{branch_name}" if remote else branch_name) in found_branches
+    def does_branch_exist(self, branch_name: str) -> bool:
```
Comment (Jorg88) [inline src/mpyl/utilities/repo/__init__.py:316]: I can't check from my phone, but I thought this function was also used to check if local branches exist?

File: tests/artifacts/test_artifacts.py [hunk 6]
```diff
@@ -1,74 +1,36 @@
...
 
-                assert self.tmp_build_artifacts.exists()
-            finally:  # cleanup
-                shutil.rmtree(self.tmp_folder, ignore_errors=True)
-                if artifact_repo:
-                    artifact_repo.delete_remote_branch(branch_name=self.test_branch)
-                os.chdir(old_working_dir)
+        artifacts.pull(
+            artifact_repo_config=artifact_repo_config, branch="test/build-artifacts"
+        )
```
Comment (Jorg88) [inline tests/artifacts/test_artifacts.py:33]: We don't want to do any assertions the files actually are pushed and pulled?

File: src/mpyl/artifacts/build_artifacts.py (hunk 2 above)
Comment (SamTheisens) [reply src/mpyl/artifacts/build_artifacts.py:48]: Yeah, good point. I think it's useful to have a target folder within the artifact repo for other reasons as well. In case of argocd you may have different pr/acc/prod folders in the same repo

File: src/mpyl/artifacts/build_artifacts.py (hunk 3 above)
Comment (SamTheisens) [reply src/mpyl/artifacts/build_artifacts.py]: Yes, we only need to take the caching scenario into account, deploy artifacts are never popped.

File: src/mpyl/artifacts/build_artifacts.py (hunk 4 above)
Comment (SamTheisens) [reply src/mpyl/artifacts/build_artifacts.py]: Nothing, it's better. This was just a test.

File: tests/artifacts/test_artifacts.py (hunk 6 above)
Comment (SamTheisens) [reply tests/artifacts/test_artifacts.py:76]: I'm not sure that's useful with the test being ignored?

File: src/mpyl/cli/build.py [hunk 7]
```diff
@@ -442,5 +450,93 @@ def clean(obj: CliContext, filter_):
         obj.console.print("Nothing to clean")
 
 
+@build.command(help="Pull build artifacts from remote artifact repository")
```
[the remark this answers is not in the document]
Comment (SamTheisens) [reply src/mpyl/cli/build.py:453]: I agree, let me see if we can nest another level deeper.

File: src/mpyl/utilities/repo/__init__.py [hunk 8]
```diff
@@ -116,6 +132,37 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            raise ValueError("Cannot clone repository without credentials")
+
+        repo = Repo.clone_from(
+            url=creds.to_url_with_credentials,
+            to_path=repo_path,
+        )
+        user_name = creds.user_name
+        with repo.config_writer() as writer:
+            writer.set_value("user", "name", user_name)
+            writer.set_value("user", "email", creds.email or "somebody@somwhere.com")
```
[the remark this answers is not in the document]
Comment (SamTheisens) [reply src/mpyl/utilities/repo/__init__.py:148]: It's already falling back to `somebody@somwhere.com`, or are you suggesting to make it a required field?

[the remark this answers is not in the document]
Comment (koeves) [reply src/mpyl/utilities/repo/__init__.py:148]: Yes I think we should make it required

File: tests/test_resources/artifacts_test/test_project.yml
Comment (Jorg88) [inline tests/test_resources/artifacts_test/test_project.yml:1]: did you mean to add this empty file?

**Review by Jorg88 — CHANGES_REQUESTED** [review]

File: mpyl-example.py [hunk 9]
```diff
@@ -10,13 +10,13 @@ def main(log: Logger, args: argparse.Namespace):
         from src.mpyl.steps.models import RunProperties
         from src.mpyl.utilities.pyaml_env import parse_config
         from src.mpyl.cli import MpylCliParameters
-        from src.mpyl.build import run_mpyl
+        from src.mpyl.cli.commands.build.build import run_mpyl
```
[the remark this answers is not in the document]
Comment (SamTheisens) [reply mpyl-example.py:13]: Hmm, why did we do that 🤔 . All the other command execution code is there too.

[the remark this answers is not in the document]
Comment (Jorg88) [reply mpyl-example.py:13]: Because these functions are not only used in the cli, plus we thought the 'main' mpyl function (`run_mpyl`) shouldn't be hidden so far into a subfolder.

**Review by Jorg88 — APPROVED** [review]
There is still the empty file `tests/test_resources/artifacts_test/test_project.yml`, but looks good otherwise!

**Review by koeves — APPROVED** [review]

File: src/mpyl/utilities/repo/__init__.py [hunk 10]
```diff
@@ -116,6 +132,37 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
         return self
 
+    @staticmethod
+    def from_clone(config: RepoConfig, repo_path: Path):
+        creds = config.repo_credentials
+        if not creds:
+            raise ValueError("Cannot clone repository without credentials")
+
+        repo = Repo.clone_from(
+            url=creds.to_url_with_credentials,
```
Comment (Jorg88) [inline src/mpyl/utilities/repo/__init__.py:142]: Actually I have one more question, why did you change this to use the username/password way instead of just the ssh url? It even warns that the support for this has been dropped on August 13, 2021

Comment (SamTheisens) [reply src/mpyl/utilities/repo/__init__.py:142]: Ah yes. The reason I used the ssh URL is that it was a way to get it to work in github actions: https://github.com/SamTheisens/mpyl-example-gha/pull/6/files#diff-ee8dae421069ba04c12f6746ff297be2674d2dbd629e36c374e5f07245fdec71R11-R14
I don't have ssh credentials there. 

But if you need ssh access on Jenkins, I think you need to add that feature. I can't really test that myself because I don't have access to Jenkins. Would be good if we keep the possibility to use either.

**Review by koeves — APPROVED** [review]
